### ISSUE TITLE
functoolz.memoize as class rather than closure

### ISSUE THREAD

**Issue by justanr** [pytoolz/toolz#231]
It was mentioned [here](https://github.com/pytoolz/toolz/pull/230#issuecomment-77939660) that memoize may be better served as a class rather than a closure.

The challenges of writing a class based decorator are far different than writing a functional closure decorator. Not so much greater that it isn't worth it; rather, a class based decorator is far more flexible than a closure decorator could hope.

An obvious example here is the ability to easily expose the memoized callable, the cache and the cache key. In addition, I've added a method to invalidate part or all of the cache -- it's left to the user to determine _when_ the cache invalidation happens.

However, exactly how flexible should the decorator be? I think, in order of importance, it should be able to memoize:
- functions -- easy, already working
- methods -- again, easy and already working
- classes -- mostly easy, memoizes `__init__` and short circuits the cache `has_kwargs` check to True
- descriptors -- things get real murky here and I'm not 100% sure it's worth it.

There are other things I'm pondering as well.
- Should it be hashable? My gut says no, caches change. If we hash the cached callable and then have a miss, the hash changes.
- Should it be pickleable? I'm not sure. If it should be, does the cache get pickled as well?

The working class is [here](https://github.com/justanr/toolz/blob/memo_as_class/toolz/functoolz.py#L318). Thoughts?

**Comment by mrocklin** [comment]
I think that making `memoize` in to a class seems reasonable.

I personally wouldn't worry about making it hashable, pickleable, or even eq comparable, but that's just me.

**Comment by justanr** [comment]
How about caching descriptors  -- things like `@classmethod` or `@staticmethod`, not to mention whatever a user will come up with? It's not terribly difficult, but it also isn't pretty.

**Comment by mrocklin** [comment]
I don't have a good idea of how difficult this would be, though obviously my preference would be to "support all of the things" as long as I'm not the one doing the work :)

@eriknw might also have thoughts on this approach.

**Comment by eriknw** [comment]
There's some good stuff in here.

I wouldn't worry about the `invalidate` method (which means you can also remove `__check_cache`).  By exposing `cache`, the user can modify it as they wish.  I would also stick with the argument names `key`, `cache`, and `func`.  Making these attributes read-only (as they are now by using `@property`) seems like a fine choice, and may be a convention we should adopt more fully.

Regarding memoizing things like `@classmethod`, I think your description in the docstring is sufficient.  `memoize` should wrap a "normal" function, and `@classmethod` needs to be the outer-most decorator.

**Comment by eriknw** [comment]
Regarding hashing, I don't think we need to explicitly define `__hash__` method.  Use the default hash.

Regarding pickling, we now need to make a choice:
  1. explicitly disable pickling
  2. pickle `func` and `key`
  3. pickle everything including the `cache`

Previously, pickling wasn't supported with the closure implementation.  I would favor (1) or (2) above, with a preference to (1).

**Comment by mrocklin** [comment]
I personally prefer 1 with an informative message.

**Comment by justanr** [comment]
@mrocklin Wrapping a descriptor *properly* isn't fun, see [wrapt](https://github.com/GrahamDumpleton/wrapt/) for a very indepth guide. Of course, there's always the easy way of 'if hasattr(wrapped, '__get__'): ...` but that feels more like hitting it with a rock until it acquiesces.

@eriknw I'm fine with invalidate disappearing, however by keeping `memoize.__check_cache` around, it's possible to subtype without needlessly reimplementing it. For example, I had an eye towards making an LRU using OrderedDict as the cache. That said, inheriting from memoize is...tricky because it actually ends up as an instance of curry, but that's another matter.

I'll change the `memoize.wrapped` property back to `memoize.func` but I feel the name is slightly misleading. However, given this a library aimed at bringing *functional* features to Python, I can see `func` being the canonical name. Is the error message, "memoized objects are non-Pickeable" sufficient enough?

As for making `memoize.cache` a property, I'm not 100% sold on this. The only thing it prevents is replacing `memoize.cache` with a completely different thing, but anyone who would intentionally do that, will just as likely turn to `memoize._cache`  -- or worse: `memozie(...).__dict__['_cache']`. Thoughts?

There's no love lost on my end for disabling pickling. Should we raise a TypeError, PicklingError (seems to be the preferred pickle way) or another error?

**Comment by mrocklin** [comment]
I personally don't mind leaving fields as public.  

In regards to LRU my vote would be to make an lru `MutableMapping` object and then feed that in through the cache argument rather than subclassing `memoize`.  This seems like a slightly more orthogonal mode of composition.  I could also use an lru dict in other projects unrelated to memoization.

**Comment by justanr** [comment]
@mrocklin Consider...

    @curry
    class lru_memoize(memoize.func): # decorated classes are hard
        def __init__(self, func, cache=None, key=None, size=float('inf')):
            super(lru_memoize, self).__init__(func, cache, key)
            self.cache = OrderedDict(self.cache) # pass in cache=OrderedDict(...) if needed
            self.size = size

        def __call__(self, *args, **kwargs):
            in_cache, key = self._check_cache(*args, **kwargs)
            if in_cache:
                self.cache.move_to_last(key)
            else:
                self.cache[key] = self.func(*args, **kwargs)
                if len(self.cache) > self.size:
                    self.cache.popitem(0)
            return self.cache[key]

Also, writing this example out lead me to realize that `memoize.__call__` only needs to check `if not in_cache:` and that `memoize.__check_cache` should actually be `memoize._check_cache` (if it's to be kept)

**Comment by mrocklin** [comment]
Cool that it's that easy.  I do think that `SizedOrderedDict` would be handy on its own though.  More broadly I'd love to see lots of things implement the `MutableMapping` interface.  

I've found recent experiments with [chest](https://github.com/ContinuumIO/chest) (a spill-to-disk dictionary like `shelve`) very fruitful in multiple projects.  The fact that we can combine chest with `toolz.memoize` is just a pleasant side effect and not something we had to explicitly work towards or maintain.

In OO speak, this is a bit like the difference between subclassing and [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(object-oriented_programming)).  In functional parlance this is trying to find orthogonal composable abstractions.
